device-manager: Added secure guest capacity device plugin

zhencliu opened the pull request:
Added device plugin for SEV-SNP secure guest capacity by reading /sys/fs/cgroup/misc.capacity. This enables resource housekeeping to prevent over-scheduling of confidential VMs on worker nodes.

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as a draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
#### Before this PR:
The capacity of starting secure guests per worker node is limited, currently kubevirt doesn't consider it when starting the secure guests.

#### After this PR:
 There are secure guest capacity resources exposed by the worker nodes, which determines how many confidential VMs can be created in the node. The scheduler consumes one for each secure guest. The number of the secure guests started per node is limited.  This is applicable to SNP and the available resources are available under `/sys/fs/cgroup/misc.capacity`.

### References
<!-- optional,
  Use `Fixes #<issue number>(, Fixes #<issue_number>, ...)` format, to close the issue(s) when PR gets merged.
  Use `Partially addresses #<issue number>` to link an issue without closing it when the PR merges.
  
- Fixes #
- Partially addresses #
-->
<!-- optional,
  VEP tracking issue if this PR is implementing one.
  For additional info about VEP tracking issue, see https://github.com/kubevirt/enhancements#process
  
- VEP tracking issue: https://github.com/kubevirt/enhancements/issues/<vep_tracking_issue_number>
-->

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

rmohr commented on pkg/virt-handler/device-manager/secure_guest_capacity_device.go:
```diff
@@ -0,0 +1,296 @@
+	CapacityPath = "/sys/fs/cgroup/misc.capacity"
+
+	// Use the same resource names as CoCo's extended resources
+	TDXResourceName = "tdx.intel.com/keys"
```
Would you consider `kubevirt.io` as prefix? While others could use them as well when virt-handler is available, we still have to use our own namespace to avoid conflicts and make ownership clear.

Something like `kubevirt.io/confidential-compute/tdx` an `kubeviort.io/confidential-compute/sev` would be good names.

rmohr replied:
So far we treat nfd as optional. If we want to expose it, we have add a prefix, that's the k8s convention. If we want to integrate with nfd in addition, I would suggest KubeVirt config parameters to allow reusing nfd, where you then would skip exposing this resource via kubevirt.

vladikr replied:
@mythi, is there a reason why they are not using a device plugin? Would kata be open to developing a shared device plugin so that we can avoid collisions?

zhencliu replied:
Is it possible to extend the KubeVirt CR spec to support this configuration?

```
configuration:
    confidentialCompute:
      extendedResourceCapacity: true
```

When extendedResourceCapacity is true: KubeVirt relies on NFD/extended resource to advertise the resource and simply consumes it.
When extendedResourceCapacity is false (Default): KubeVirt manages the device plugin internally to advertise and consume the resource.

Explicit configuration of the resource names may be not necessary, as the names are constant in both the NFD and device plugin scenarios.

If extendedResourceCapacity is set to true but the resource is not advertised (e.g., if NFD is removed), this should be treated as a configuration error.

zhencliu replied:
> Although I'm not sure a device plugin resource resource can work as a pod overhead because device allocations are coupled with containers, not pods. It should be easy to test).

I tested this:
`kubectl patch runtimeclass "kata-qemu-tdx" --type=merge -p='{"overhead":{"podFixed":{"tdx.intel.com/keys":1}}}'`
Started a pod with kata-qemu-tdx runtime class, it worked.
```
kubectl describe nodes/tdx-worker|grep tdx
                    kubevirt.io/tdx=true
  tdx.intel.com/keys:             15
  tdx.intel.com/keys:             15
  tdx.intel.com/keys             1            0

```

rmohr reviewed (commented):
Nice. Two suggestions.

rmohr commented on pkg/virt-handler/device-manager/secure_guest_capacity_device.go:
```diff
@@ -0,0 +1,296 @@
+ *
+ */
+
+package device_manager
```
I am not sure if we finally limit each virt-handler to the node it is running on. We may. If so, it would be an option.

rmohr commented on pkg/virt-handler/device-manager/generic_device.go:
```diff
@@ -59,11 +59,17 @@ func NewGenericDevicePlugin(deviceName string, devicePath string, maxDevices int
+		// No device is monitored, e.g. secure guest capacity
+		initHealth = pluginapi.Healthy
+	}
+
```
If we have a very stable framework no issue with that, but the kubevirt device plugin (while you have a PR with nice additional improvements) handles all the quirks with kubelet very well. Therefore my suggestion.

mythi commented on pkg/virt-controller/services/template.go:
```diff
@@ -1563,6 +1567,8 @@ func (t *TemplateService) VMIResourcePredicates(vmi *v1.VirtualMachineInstance,
 				return t.clusterConfig.HostDevicesWithDRAEnabled() && isHostDevVMIDRA(vmi)
 			}, WithHostDevicesDRA(vmi.Spec.Domain.Devices.HostDevices)),
 			NewVMIResourceRule(util.IsSEVVMI, WithSEV()),
+			NewVMIResourceRule(util.IsTDXVMI, WithTDXCapacity(t.clusterConfig.SecureGuestCapacityExtendedResourceEnabled())),
```
This is not a kata resource. Kata runtimeclasses are only consumers of someone else registering the resource.

Commit:
device-manager: Added secure guest capacity device plugin

Added device plugin for AMD SEV-SNP secure guest capacity by reading /sys/fs/cgroup/misc.capacity.
This enables resource housekeeping to prevent over-scheduling of confidential VMs on worker nodes.
Note:
1. SEV/SEV-ES are not considered because they are experimental and SEV-SNP is the target
2. Intel TDX capacity is implemented in a different way, and covered by another PR

Signed-off-by: Zhenchao Liu <zhencliu@redhat.com>

Aseeef commented:
@zhencliu it seems the GetSecureGuestCapacity is going to be duplicated between our PRs (I have a similar GetMiscCapacity). How do you propose we converge on one implementation?

zhencliu commented:
Hi, due to the recent layoffs affecting Red Hat engineers in China, I cannot work on this, @alancaldelas will take over, thanks.

jean-edouard commented:
> Hi, due to the recent layoffs affecting Red Hat engineers in China, I cannot work on this, @alancaldelas will take over, thanks.

Thank you so much @zhencliu for all your hard work around confidential computing.
I'm so sorry to hear you were impacted by this... I wish you the best of luck for whatever comes next.

MatiasVara commented:
is this work required for https://github.com/kubevirt/enhancements/issues/80 to get graduation to beta?

alaypatel07 commented:
/cc @alaypatel07

alancaldelas commented:
> is this work required for https://github.com/kubevirt/enhancements/issues/80 to get graduation to beta?

For SNP I wouldn't think so, we have e2e testing for SNP isn't that the requirement?

alancaldelas commented:
Taking this over per @zhencliu's note above. Opened #17638 with the commit rebased onto current main (one merge conflict resolved in `pkg/virt-handler/cgroup/cgroup_test.go`); zhencliu's authorship is preserved on the commit. Could a maintainer please `/close` this PR in favor of #17638? Thanks.

EdDev commented:
Superseded by https://github.com/kubevirt/kubevirt/pull/17638 
/close
